src/persist-client/examples/maelstrom/txn.rs

```diff
@@ -0,0 +1,501 @@
+            let v = v.map_err(|err| MaelstromError {
+                code: ErrorCode::Crash,
+                text: format!("invalid val {}", err),
+            })?;
+            if ret.contains_key(&k) {
+                return Err(MaelstromError {
+                    code: ErrorCode::Crash,
+                    text: format!("unexpected duplicate key {:?}", k),
+                });
+            }
+            ret.insert(k, v);
+        }
+        Ok(ret)
+    }
+
+    fn eval_txn(
+        state: &HashMap<MaelstromKey, MaelstromVal>,
+        req_ops: &[ReqTxnOp],
+    ) -> (Vec<(MaelstromKey, MaelstromVal, i64)>, Vec<ResTxnOp>) {
+        let mut res_ops = Vec::new();
+        let mut updates = Vec::new();
+        let mut txn_state = HashMap::new();
+
+        for req_op in req_ops.iter() {
+            match req_op {
+                ReqTxnOp::Read { key } => {
+                    let current = txn_state
+                        .get(&MaelstromKey(*key))
+                        .or_else(|| state.get(&MaelstromKey(*key)));
+                    let val = current.cloned().unwrap_or_default().0;
+                    res_ops.push(ResTxnOp::Read { key: *key, val })
+                }
+                ReqTxnOp::Append { key, val } => {
+                    let current = txn_state
+                        .get(&MaelstromKey(*key))
+                        .or_else(|| state.get(&MaelstromKey(*key)));
+                    let mut vals = match current {
+                        Some(val) => {
+                            // Retract the value we're about to overwrite.
+                            updates.push((MaelstromKey(*key), val.clone(), -1));
+                            val.clone()
+                        }
+                        None => MaelstromVal::default(),
+                    };
+                    vals.0.push(*val);
+                    txn_state.insert(MaelstromKey(*key), vals.clone());
+                    updates.push((MaelstromKey(*key), vals, 1));
+                    res_ops.push(ResTxnOp::Append {
+                        key: key.clone(),
+                        val: *val,
+                    })
+                }
+            }
+        }
+
+        debug!(
+            "eval_txn\n  req={:?}\n  res={:?}\n  updates={:?}\n  state={:?}\n  txn_state={:?}",
+            req_ops, res_ops, updates, state, txn_state
+        );
+        (updates, res_ops)
+    }
+
+    async fn advance_since(&mut self) {
+        // To keep things interesting, advance the since.
+        const SINCE_LAG: u64 = 10;
+        let new_since = self.read_ts.saturating_sub(SINCE_LAG);
+        debug!("downgrading since to {}", new_since);
+        self.read
+            .downgrade_since(Antichain::from_elem(new_since))
+            .await;
+        self.since_ts = new_since;
+    }
+
+    fn extract_ts<T: TotalOrder + Copy>(frontier: &Antichain<T>) -> Result<T, MaelstromError> {
+        frontier.as_option().copied().ok_or_else(|| MaelstromError {
```

Comment:
TIL! `Antichain::as_option()` (only for totally ordered times, of course) :partying_face: